### PR TITLE
end_user/refactor/registration

### PR DESCRIPTION
**Pull request by mosayed01**
## Subtasks
- [x] Refactor registration to match new Figma design.
- [x] Add 4 TODOs need to remove in next deploy.
- [x] Refactor Phone Validation 
- [x] Add Address 

## Demo
https://youtu.be/0AsxNvqKEDE

### CODE REVIEW

**Review by ay0o5h — APPROVED** [review]

**Review by ahmednasserzaza — APPROVED** [review]

**Review by HaidyAbuGom3a — APPROVED** [review]

File: client_end_user/shared/src/commonMain/kotlin/domain/usecase/validation/IValidationUseCase.kt [hunk 1]
```diff
@@ -46,26 +49,40 @@ class ValidationUseCaseUseCase : IValidationUseCase {
         }
     }
 
-    override fun validatePhone(phone: String, currency: String) {
+    override fun validateAddress(address: String) {
+        // todo: add address validation regex like this one: ^[A-Za-z0-9\s.,-]+$
+        if (address.isEmpty()) {
```
Comment (Saif-Alhaider) [inline client_end_user/shared/src/commonMain/kotlin/domain/usecase/validation/IValidationUseCase.kt:54]:
```suggestion
        if (address.isBlank()) {
```

File: client_end_user/shared/src/commonMain/kotlin/domain/usecase/validation/IValidationUseCase.kt [hunk 2]
```diff
@@ -46,26 +49,40 @@ class ValidationUseCaseUseCase : IValidationUseCase {
...
+    }
+
+    override fun validatePhone(phone: String, currency: String?) {
+        if (currency == null) {
+            if (phoneRegexMap.values.any { it.matches(phone) }) {
+                return
+            } else {
+                throw AuthorizationException.InvalidPhoneException
+            }
+        }
```
Comment (Saif-Alhaider) [inline client_end_user/shared/src/commonMain/kotlin/domain/usecase/validation/IValidationUseCase.kt:66]:
```suggestion
if (currency == null) {
    phoneRegexMap.values.any { it.matches(phone) }.takeIf { it } ?: throw AuthorizationException.InvalidPhoneException
}
```

File: client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registration/RegistrationScreenEffect.kt [hunk 3]
```diff
@@ -1,7 +1,11 @@
 package presentation.auth.signup.registration
 
 sealed interface RegistrationScreenEffect {
-    data class NavigateToSubmitRegistrationScreen(val username: String, val password: String) :
-        RegistrationScreenEffect
+    data class NavigateToSubmitRegistrationScreen(
+        val username: String,
+        val email: String,
+        val password: String
+    ) : RegistrationScreenEffect
```
Comment (Saif-Alhaider) [inline client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registration/RegistrationScreenEffect.kt:8]: why this data class has variables that are not related to its name ?
also the class name is more like a verb not a noun could make it more like a noun when adding "on" before it
"OnNavigateToSubmitRegistrationScreen"

File: client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registration/RegistrationScreenModel.kt [hunk 4]
```diff
@@ -20,6 +20,11 @@ class RegistrationScreenModel(private val validation: IValidationUseCase)
         }
     }
 
+    override fun onEmailChanged(email: String) {
+        updateState { it.copy(email = email) }
+        tryCatch { validation.validateEmail(email); clearErrors() }
```
Comment (Saif-Alhaider) [inline client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registration/RegistrationScreenModel.kt:25]: making it multiple lines instead of single line is more readable

File: client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registration/RegistrationUIState.kt [hunk 5]
```diff
@@ -2,9 +2,11 @@ package presentation.auth.signup.registration
 
 data class RegistrationUIState(
     val username: String = "",
+    val email: String = "",
     val password: String = "",
     val isLoading: Boolean = false,
     val isUsernameError: Boolean = false,
+    val isEmailError: Boolean = false,
     val isPasswordError: Boolean = false,
```
Comment (Saif-Alhaider) [inline client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registration/RegistrationUIState.kt:10]: could be seperated in another data class like RegisterValidation or ValidationErrorState or could make a data class for Textfield and that textfield has type and error state...etc

File: client_end_user/shared/src/commonMain/kotlin/domain/usecase/ManageAuthunticationUseCase.kt [hunk 6]
```diff
@@ -31,13 +33,15 @@ class ManageAuthenticationUseCase(
...
+        phone: String,
+        address: String,
     ): Boolean {
         with(validation) {
             validateFullName(fullName); validateUsername(username); validatePassword(password)
-            validateEmail(email)
+            validateEmail(email); validatePhone(phone); validateAddress(address)
         }
-        return remoteGateway.createUser(fullName, username, password, email).name.isNotEmpty()
+        return remoteGateway.createUser(fullName, username, password, email, phone, address).name.isNotEmpty()
```
Comment (Saif-Alhaider) [inline client_end_user/shared/src/commonMain/kotlin/domain/usecase/ManageAuthunticationUseCase.kt:44]: make data class instead of passing 6 variables could be like "UserInfo"

File: client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registrationSubmit/RegistrationSubmitScreenModel.kt [hunk 7]
```diff
@@ -60,21 +57,21 @@ class RegistrationSubmitScreenModel(
...
     override fun onPhoneChanged(phone: String) {
         updateState { it.copy(phone = phone) }
         tryCatch { validation.validatePhone(phone,_language.value.value); clearErrors() }
     }
 
+    override fun onAddressChanged(address: String) {
+        updateState { it.copy(address = address) }
+        tryCatch { validation.validateAddress(address); clearErrors() }
+    }
+
```
Comment (Saif-Alhaider) [inline client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registrationSubmit/RegistrationSubmitScreenModel.kt:69]: could add clearErrors() to the function tryCatch instead of invoking it many times
if the function tryCatch is built in the language make another function that uses it and invokes  clearErrors() at the end

File: client_end_user/shared/src/commonMain/kotlin/resources/strings/ar/ArabicEG.kt [hunk 8]
```diff
@@ -90,5 +90,7 @@ data class ArabicEG(
     override val orderNow: String = "أطلب دلوقتي",
     override val total: String = "المجموع:",
     override val yourCart: String = "العربة",
-    override val pleaseLogin: String="سجل دخول يا اسطا وشوفك  حسابك"
+    override val pleaseLogin: String="سجل دخول يا اسطا وشوفك  حسابك",
```
Comment (Saif-Alhaider) [inline client_end_user/shared/src/commonMain/kotlin/resources/strings/ar/ArabicEG.kt:93]: there is double spaces here
```suggestion
    override val pleaseLogin: String="سجل دخول يا اسطا وشوفك حسابك",
```

File: client_end_user/shared/src/commonMain/kotlin/resources/strings/en/English.kt [hunk 9]
```diff
@@ -90,5 +90,7 @@ data class English(
     override val logout: String = "Logout",
     override val address: String = "Address",
     override val save: String = "Save",
-    override val pleaseLogin: String="Please log in to access your profile"
+    override val pleaseLogin: String="Please log in to access your profile",
```
Comment (Saif-Alhaider) [inline client_end_user/shared/src/commonMain/kotlin/resources/strings/en/English.kt:93]: login is one word not two words

**Review by Saif-Alhaider — COMMENTED** [review]
great job 👏
there are a little notes i've added could you read it when ur ready 🌷

File: client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registration/RegistrationScreenEffect.kt (hunk 3 above)
Comment (mosayed01) [reply client_end_user/shared/src/commonMain/kotlin/presentation/auth/signup/registration/RegistrationScreenEffect.kt:8]: it's represent the effect itself. so it's okay to be a verb

**Review by NadaFeteiha — APPROVED** [review]
